### PR TITLE
Public key pinning

### PR DESCRIPTION
**Pull request by kikettas**
Replaced certificate pinning implementation with public key pinning, using [TrustKit](https://github.com/datatheorem/TrustKit/) this time.

### How to test
Create a SDK with `GINISDKBuilder` with the Public Key pinning config and a valid public key hash. Then do the same but with an invalid public key hash.

### Merging
Automatic.
 
### Issues
#60

### CODE REVIEW

File: Gini-iOS-SDK/GINIURLSessionDelegate.m [hunk 1]
```diff
@@ -7,67 +7,26 @@
...
-}
-
-- (instancetype)initWithNSURLSessionDelegate:(NSArray<NSString *> *)certificatePaths {
-    self = [super init];
-    if (self) {
-        _nsCertificatePaths = certificatePaths;
-    }
-    return self;
++ (instancetype)urlSessionDelegate {
+    return [self alloc];
```
Comment (nsobadjiev) [inline Gini-iOS-SDK/GINIURLSessionDelegate.m]: Why not `init`?

File: README.md [hunk 2]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
 is provided at the instance's `injector` property) to create the manager instances and to manage the dependencies
 between the managers and makes those manager instances available as properties.
 
-## Requirements
+## Public Key pinning
+The Gini iOS SDK allows you to enable public key pinning with the Gini API through [TrustKit](https://github.com/datatheorem/TrustKit/) library. In case that you want to implement it, you just need to pass a dictionary with all the parameters required into one of the `GINISDKBuilder` initializers, as follows:
```
Comment (nsobadjiev) [inline README.md]: through **the** TrustKit library

Comment (nsobadjiev) [reply README.md]: In case <strike>that</strike> you

File: README.md [hunk 3]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
...
+                        ],
+                }}};
+
+GiniSDK *sdk = [[GINISDKBuilder anonymousUserWithClientID:@"your_client_id"
+                                             clientSecret:@"your_client_secret"
+                                          userEmailDomain:@"your_user_email_domain"
+                                   publicKeyPinningConfig:trustKitConfig] build];
+```
+
+When using `TrustKit` some messages are shown in the console log, being possible to either specify a custom log for that messages or disable it (setting it as `nil`). To do so just use the `TrustKit.setLoggerBlock` method before initializing the `trustKitConfig`.
```
Comment (nsobadjiev) [inline README.md]: ...console log. It is possible to either specify a custom log for messages or disable it altogether (setting it as `nil`).

File: README.md [hunk 4]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
...
 
-- iOS 7.1+
+## Requirements
+- iOS 8.0+
```
Comment (nsobadjiev) [inline README.md]: Did you communicate that with the account managers?

File: Gini-iOS-SDK.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -646,6 +646,7 @@
 						};
 					};
 					23D53C031934BC85001A957E = {
+						ProvisioningStyle = Automatic;
```
Comment (nsobadjiev) [inline Gini-iOS-SDK.xcodeproj/project.pbxproj]: Again, you have changes in the project file that are unrelated to the feature.

File: GiniSDK Example/Assets.xcassets/AppIcon.appiconset/Contents.json [hunk 6]
```diff
@@ -84,6 +84,11 @@
       "idiom" : "ipad",
       "size" : "83.5x83.5",
       "scale" : "2x"
+    },
+    {
+      "idiom" : "ios-marketing",
+      "size" : "1024x1024",
+      "scale" : "1x"
```
Comment (nsobadjiev) [inline GiniSDK Example/Assets.xcassets/AppIcon.appiconset/Contents.json]: What is this change? I assume it was just because you opened the Assets file?

File: README.md (hunk 4 above)
Comment (kikettas) [reply README.md]: Yes, I have a list with the min version of the Apps in the App Store being iOS 8 the lower one.

File: Gini-iOS-SDK/GINIURLSessionDelegate.m (hunk 1 above)
Comment (kikettas) [reply Gini-iOS-SDK/GINIURLSessionDelegate.m]: Because in this case is the same, since it is the default initializer. For me there isn't a big difference.

File: Gini-iOS-SDK.xcodeproj/project.pbxproj (hunk 5 above)
Comment (kikettas) [reply Gini-iOS-SDK.xcodeproj/project.pbxproj]: I had to add it because for the tests weren't running. 
I want to migrate the whole project to a new structure, since it was set up with CocoaPods < 1.0 and It is less clear in my opinion (and more prone to errors).

File: GiniSDK Example/Assets.xcassets/AppIcon.appiconset/Contents.json (hunk 6 above)
Comment (kikettas) [reply GiniSDK Example/Assets.xcassets/AppIcon.appiconset/Contents.json]: Xcode did it automatically, I didn't want it for this PR but I think it is not a big deal.

**Comment by kikettas** [comment]
@nsobadjiev Yes, the API was changed because the ones that were using the certificate pinning want to use public key pinning instead.
About the conditional frameworks, I don't know a way to make them optional. Do you know something about this?

File: Gini-iOS-SDK/GINISDKBuilder.h [hunk 7]
```diff
@@ -91,19 +93,19 @@
...
  * Creates an instance of the GINISDKBuilder where anonymous users are used.
  *
- * @param clientId          The application's clientID for the Gini API.
+ * @param clientId                The application's clientID for the Gini API.
  *
- * @param emailDomain       The domain of the email address.
- * @param certificatePaths  Local certificate paths for certificate pinning.
+ * @param emailDomain             The domain of the email address.
+ * @param publicKeyPinningConfig  Public key pinning configuration. More information about available parameters can be found here
+ *                                (https://datatheorem.github.io/TrustKit/documentation/Classes/TrustKit.html#/c:objc(cs)TrustKit(py)pinningValidator)
```
Comment (a-szotyori) [inline Gini-iOS-SDK/GINISDKBuilder.h]: Why is the link here in parantheses?

File: README.md [hunk 8]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
...
+## Public Key pinning
+The Gini iOS SDK allows you to enable public key pinning with the Gini API through the [TrustKit](https://github.com/datatheorem/TrustKit/) library. In case you want to implement it, you just need to pass a dictionary with all the parameters required into one of the `GINISDKBuilder` initializers, as follows:
+
+**Swift**
+```swift
+let trustKitConfig = [
+            kTSKSwizzleNetworkDelegates: false,
+            kTSKPinnedDomains: [
+                "example.com": [
+                    kTSKExpirationDate: "2017-12-01",
```
Comment (a-szotyori) [inline README.md]: In the Android TrustKit implementation, if public key pinning is not enforced, TrustKit will do a report about public key mismatch, but it won't fail the server check, effectively disabling pinning.

Can you check in the iOS implementation for that and if the behavior is the same enable `kTSKEnforcePinning` here, too?

Comment (a-szotyori) [inline README.md]: In the Android TrustKit implementation, they send reports to to `https://overmind.datatheorem.com/trustkit/report`. It should be disabled with `kTSKDisableDefaultReportUri`. We should add that to the examples.

File: README.md [hunk 9]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
 is provided at the instance's `injector` property) to create the manager instances and to manage the dependencies
 between the managers and makes those manager instances available as properties.
 
-## Requirements
+## Public Key pinning
+The Gini iOS SDK allows you to enable public key pinning with the Gini API through the [TrustKit](https://github.com/datatheorem/TrustKit/) library. In case you want to implement it, you just need to pass a dictionary with all the parameters required into one of the `GINISDKBuilder` initializers, as follows:
+
+**Swift**
+```swift
+let trustKitConfig = [
```
Comment (a-szotyori) [inline README.md]: In the Android documentation I show a working configuration with the current *.gini.net public key and an invalid key hash I created with `$ echo "0" | openssl dgst -sha256 -binary | openssl enc -base64`. Although it will get out-dated it could be more useful now for our clients to have a valid example.

File: README.md [hunk 10]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
...
+                                       userEmailDomain: "your_user_email_domain"
+                                       publicKeyPinningConfig: trustKitConfig).build()
+
+```
+
+**Objective-C**
+
+```objective-c
+NSDictionary *trustKitConfig = @{
+kTSKPinnedDomains: @{
```
Comment (a-szotyori) [inline README.md]: Both examples should show the same configuration.

File: README.md [hunk 11]
```diff
@@ -27,9 +27,54 @@ described in the [integration guide](http://developer.gini.net/gini-sdk-ios/docs
...
-## Requirements
+## Public Key pinning
+The Gini iOS SDK allows you to enable public key pinning with the Gini API through the [TrustKit](https://github.com/datatheorem/TrustKit/) library. In case you want to implement it, you just need to pass a dictionary with all the parameters required into one of the `GINISDKBuilder` initializers, as follows:
+
+**Swift**
+```swift
+let trustKitConfig = [
+            kTSKSwizzleNetworkDelegates: false,
+            kTSKPinnedDomains: [
+                "example.com": [
```
Comment (a-szotyori) [inline README.md]: Should be `gini.net` and we have to add `kTSKIncludeSubdomains` so that pinning for `api.gini.net` and `user.gini.net` is applied.

File: README.md (hunk 8 above)
Comment (kikettas) [reply README.md]: Yes, there is that field also on iOS, I was using it to test it but I forgot to add it to the documentation. 💪

Comment (kikettas) [reply README.md]: I thought that it wasn't send if you don't specify the url. Good point!

File: README.md (hunk 9 above)
Comment (kikettas) [reply README.md]: First I had it with "gini.net", but I decided to make it more generic. You are right, It could be better to make it more specific in this case.

File: README.md (hunk 10 above)
Comment (kikettas) [reply README.md]: I copied both from TrustKit repo, thinking that it was the same in both cases. 🤦‍♂️
